### PR TITLE
Add shuffle partition and write parquet for FeatureTable

### PR DESCRIPTION
**Pull request by hkvision**

### CODE REVIEW

**Comment by jason-dai** [comment]
> Spark DataFrame write parquet can only write to hdfs or single node local?
> For other nodes, the saved results are temp task files:
> 
> ```
> (base) [root@clx008 ~]# cd /opt/work/kai/dlrm_saved_parquet/_temporary/0/
> (base) [root@clx008 0]# ls
> task_20210407210627_0053_m_000006  task_20210407210627_0053_m_000102  task_20210407210627_0053_m_000198  task_20210407210627_0053_m_000294
> task_20210407210627_0053_m_000014  task_20210407210627_0053_m_000110  task_20210407210627_0053_m_000206  task_20210407210627_0053_m_000302
> task_20210407210627_0053_m_000022  task_20210407210627_0053_m_000118  task_20210407210627_0053_m_000214  task_20210407210627_0053_m_000310
> task_20210407210627_0053_m_000030  task_20210407210627_0053_m_000126  task_20210407210627_0053_m_000222  task_20210407210627_0053_m_000318
> task_20210407210627_0053_m_000038  task_20210407210627_0053_m_000134  task_20210407210627_0053_m_000230  task_20210407210627_0053_m_000326
> task_20210407210627_0053_m_000046  task_20210407210627_0053_m_000142  task_20210407210627_0053_m_000238  task_20210407210627_0053_m_000334
> task_20210407210627_0053_m_000054  task_20210407210627_0053_m_000150  task_20210407210627_0053_m_000246  task_20210407210627_0053_m_000342
> task_20210407210627_0053_m_000062  task_20210407210627_0053_m_000158  task_20210407210627_0053_m_000254  task_20210407210627_0053_m_000350
> task_20210407210627_0053_m_000070  task_20210407210627_0053_m_000166  task_20210407210627_0053_m_000262  task_20210407210627_0053_m_000358
> task_20210407210627_0053_m_000078  task_20210407210627_0053_m_000174  task_20210407210627_0053_m_000270  task_20210407210627_0053_m_000366
> task_20210407210627_0053_m_000086  task_20210407210627_0053_m_000182  task_20210407210627_0053_m_000278  _temporary
> task_20210407210627_0053_m_000094  task_20210407210627_0053_m_000190  task_20210407210627_0053_m_000286
> ```
> 
> cc @jason-dai

Yes - I think you will pass a single path and Spark will put everything under that folder.

**Comment by hkvision** [comment]
Writing Spark df to parquet to hdfs is terribly slow, only read parquet from NFS and write to hdfs takes ~2000s.
Would try other ways.

**Comment by hkvision** [comment]
[WIP] Performance results on AEP standalone cluster of 8 nodes:
- Original https://github.com/intel-analytics/analytics-zoo/blob/master/pyzoo/zoo/examples/friesian/feature/dlrm_preprocessing.py: ~540s (sometimes 563s or 517s, may need several runs to get stable)

**Comment by hkvision** [comment]
@jason-dai Take a look at the results above. Any comments or ideas? Need to implement a rdd.map save to local in Scala?

File: zoo/src/main/scala/com/intel/analytics/zoo/friesian/python/PythonFriesian.scala [hunk 1]
```diff
@@ -180,4 +184,63 @@ class PythonFriesian[T: ClassTag](implicit ev: TensorNumeric[T]) extends PythonZ
...
+
+  def dfWriteParquet(df: DataFrame, path: String, mode: String = "overwrite"): Unit = {
+    df.write.mode(mode).parquet(path)
+  }
+
+  // TODO: can convert schema to case class.
+  // https://gist.github.com/yoyama/ce83f688717719fc8ca145c3b3ff43fd
+  case class Record(y: Int, xInt: List[Double], xCat: List[Int])
+
+  def dfSaveParquet(df: DataFrame, path: String, mode: String = "overwrite", partSize: Int = 100000):
```
Comment (hkvision) [inline zoo/src/main/scala/com/intel/analytics/zoo/friesian/python/PythonFriesian.scala:206]: @jason-dai For large subpartition size, it is extremely slow. For smaller size, it is faster, but still slower than save to numpy. See the issue for my test results.

**Comment by hkvision** [comment]
Error:(201, 14) inferred type arguments [org.apache.spark.sql.Row,Array] do not conform to method shuffle's type parameter bounds [T,CC[X] <: TraversableOnce[X]]
Error occurred in an application involving default arguments.
      Random.shuffle(iter.toArray).toIterator

Error:(201, 27) type mismatch;
 found   : Array[org.apache.spark.sql.Row]
 required: CC[T]
Error occurred in an application involving default arguments.
      Random.shuffle(iter.toArray).toIterator

@jason-dai If input is an array, Random.shuffle would report compilation error...

**Comment by jason-dai** [comment]
Try 

> Error:(201, 14) inferred type arguments [org.apache.spark.sql.Row,Array] do not conform to method shuffle's type parameter bounds [T,CC[X] <: TraversableOnce[X]]
> Error occurred in an application involving default arguments.
> Random.shuffle(iter.toArray).toIterator
> 
> Error:(201, 27) type mismatch;
> found : Array[org.apache.spark.sql.Row]
> required: CC[T]
> Error occurred in an application involving default arguments.
> Random.shuffle(iter.toArray).toIterator
> 
> @jason-dai If input is an array, Random.shuffle would report compilation error...

You can use `toSeq`

**Comment by hkvision** [comment]
Oh yes, toSeq works. Sorry about that...

**Comment by hkvision** [comment]
dlrm_preprocessing2.py uses sortWithInPartitions for shuffle and save to hdfs with 1 replication.

**Comment by hkvision** [comment]
This PR is not for merge. Just to trace some previous work.

**Pull request closed without merging**